dmutils/request_id.py

```diff
@@ -1,50 +1,90 @@
+        return self._trace_id
+
+    @property
+    def span_id(self):
+        if not hasattr(self, "_span_id"):
+            self._span_id = self._get_first_header(current_app.config['DM_SPAN_ID_HEADERS'])
+        return self._span_id
+
+    @property
+    def parent_span_id(self):
+        if not hasattr(self, "_parent_span_id"):
+            self._parent_span_id = self._get_first_header(current_app.config['DM_PARENT_SPAN_ID_HEADERS'])
+        return self._parent_span_id
+
+    def _get_first_header(self, header_names):
+        """
+        Returns value of request's first present (and Truthy) header from header_names
+        """
+        for header_name in header_names:
+            if header_name in self.headers and self.headers[header_name]:
+                return self.headers[header_name]
         else:
-            return str(uuid.uuid4())
+            return None
+
+    def get_onwards_request_headers(self):
+        """
+            Headers to add to any further (internal) http api requests we perform if we want that request to be
+            considered part of this "trace id"
+        """
+        return dict(chain(
+            (
+                (header_name, self.trace_id)
+                for header_name in current_app.config['DM_TRACE_ID_HEADERS']
+                if self.trace_id
+            ),
+            (
+                (header_name, self.span_id)
+                for header_name in current_app.config['DM_SPAN_ID_HEADERS']
+                if self.span_id
+            ),
+        ))
 
 
 class ResponseHeaderMiddleware(object):
-    def __init__(self, app, request_id_header):
+    def __init__(self, app, trace_id_headers):
         self.app = app
-        self.request_id_header = request_id_header
+        self.trace_id_headers = trace_id_headers
 
     def __call__(self, environ, start_response):
         def rewrite_response_headers(status, headers, exc_info=None):
-            if self.request_id_header not in dict(headers).keys():
-                headers = headers + [(
-                    self.request_id_header,
-                    str(request.request_id)
-                )]
+            headers = headers + [
+                (header_name, str(request.request_id),)
+                for header_name in self.trace_id_headers
+                if header_name not in dict(headers)
+            ]
 
             return start_response(status, headers, exc_info)
 
         return self.app(environ, rewrite_response_headers)
 
 
 def init_app(app):
-    app.config.setdefault('DM_REQUEST_ID_HEADER', 'DM-Request-ID')
-    app.config.setdefault('DM_DOWNSTREAM_REQUEST_ID_HEADER', '')
+    app.config.setdefault("DM_TRACE_ID_HEADERS", (
+        (app.config.get("DM_REQUEST_ID_HEADER") or "DM-Request-ID"),
+        (app.config.get("DM_DOWNSTREAM_REQUEST_ID_HEADER") or "X-B3-TraceId"),
+    ))
+    app.config.setdefault("DM_SPAN_ID_HEADERS", ("X-B3-SpanId",))
```

Comment:
I'm not sure that these config names should be prefixed 'DM_', considering they are actually referring to a standard. Although: why is it a list? Is there any need for more than one?

Reply:
I would argue that they do because they're implemented by our own shizzle. It's the signifier that you should looking for information about this setting in our own code rather than the wider internet. Used for a similar reason I prefixed our search index properties with `dm`.

Reply:
Why multiple? Well, we need this feature for the switchover period (at least) to be able to cope with the old and new ways together. It's generally a useful feature when you have a feature which may need to work with headers specified in a different way. It would give an ability to transition to a yet another header name in the future.

Reply:
Fair enough being a list then. I can see this being more useful generally and so wanting to be a separate library, maybe even owned by the PaaS team, but sure `DM_` it is for now anyway.